### PR TITLE
feat(57028): Support inlining variables that are used in object literal shorthand

### LINKED ISSUE #57028: Support inlining variables that are used in object literal shorthand
### 🔎 Search Terms

inline, refactor, refactoring

### 🕗 Version & Regression Information

Not a recession 

### ⏯ Playground Link

_No response_

### 💻 Code

```ts
const abc = 123;
const obj = { abc };
```

Try inlining `abc`

### 🙁 Actual behavior

No refactoring offered

### 🙂 Expected behavior

We should show a `inline` refactoring that results in:

```ts
const obj = { abc: 123 };
```

### Additional information about the issue

_No response_

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #57028

### CODE REVIEW

**Review by DanielRosenwasser — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
@typescript-bot test tsserver top120

File: src/services/refactors/inlineVariable.ts [hunk 1]
```diff
@@ -152,7 +152,7 @@ function getInliningInfo(file: SourceFile, startPosition: number, tryWithReferen
     }
 
     // Try finding the declaration and nodes to replace via the reference token.
-    if (tryWithReferenceToken) {
+    if (tryWithReferenceToken || isShorthandPropertyAssignment(parent)) {
```
Comment (MariaSolOs) [inline src/services/refactors/inlineVariable.ts:155]: Why is this condition modified? The purpose of `tryWithReferenceToken` is for a lightbulb to show up only in the declaration site, and to require explicit refactor requests in reference sites.

File: src/services/refactors/inlineVariable.ts [hunk 2]
```diff
@@ -191,7 +191,7 @@ function getReferenceNodes(declaration: InitializedVariableDeclaration, checker:
     const cannotInline = FindAllReferences.Core.eachSymbolReferenceInFile(declaration.name as Identifier, checker, file, ref => {
         // Only inline if all references are reads. Else we might end up with an invalid scenario like:
         // const y = x++ + 1 -> const y = 2++ + 1
-        if (FindAllReferences.isWriteAccessForReference(ref)) {
+        if (FindAllReferences.isWriteAccessForReference(ref) && !isShorthandPropertyAssignment(ref.parent)) {
```
Comment (MariaSolOs) [inline src/services/refactors/inlineVariable.ts:194]: If changing this condition, consider updating the comment as well.

File: src/services/refactors/inlineVariable.ts (hunk 1 above)
Comment (a-tarasyuk) [reply src/services/refactors/inlineVariable.ts:155]: Should _shorthand properties_ only be allowed to change with the `invoked` trigger reason?

Comment (MariaSolOs) [reply src/services/refactors/inlineVariable.ts:155]: I believe so. Thoughts @DanielRosenwasser?

Comment (DanielRosenwasser) [reply src/services/refactors/inlineVariable.ts:155]: That sounds right. The shorthand property is effectively considered a usage. So it should work similarly to the following case

```ts
let x = 123;
let y = {
  x: x/**/
};
```

Unless there's something you're noticing that we're missing @a-tarasyuk

File: src/services/refactors/inlineVariable.ts (hunk 2 above)
Comment (MariaSolOs) [reply src/services/refactors/inlineVariable.ts:195]: Could you please explain why this change was needed? Does `FindAllReferences.isWriteAccessForReference` return true for `x`...

```ts
const x = 0;
const y = { x // ...here? }
```

File: src/services/refactors/inlineVariable.ts [hunk 3]
```diff
@@ -244,5 +245,9 @@ function getReplacementExpression(reference: Node, replacement: Expression): Exp
         return factory.createParenthesizedExpression(replacement);
     }
 
+    if (isIdentifier(reference) && isShorthandPropertyAssignment(parent)) {
```
Comment (MariaSolOs) [inline src/services/refactors/inlineVariable.ts:248]: To keep the nice documentation convention in this file, a comment here would be nice ;)

File: src/services/refactors/inlineVariable.ts (hunk 2 above)
Comment (a-tarasyuk) [reply src/services/refactors/inlineVariable.ts:195]: `ShorthandPropertyAssignment` is considered by `isWriteAccessForReference` as a write access reference.

https://github.com/microsoft/TypeScript/blob/6add96fd18d126d1505de5d96f0df6a86e5f2c21/src/services/findAllReferences.ts#L899-L902

https://github.com/microsoft/TypeScript/blob/6add96fd18d126d1505de5d96f0df6a86e5f2c21/src/services/findAllReferences.ts#L949

**Review by MariaSolOs — APPROVED** [review]

**Pull request merged**